swarm/network/stream/common_test.go

```diff
@@ -0,0 +1,151 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package stream
+
+import (
+	"errors"
+	"flag"
+	"io/ioutil"
+	"os"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	p2ptest "github.com/ethereum/go-ethereum/p2p/testing"
+	"github.com/ethereum/go-ethereum/swarm/network"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+var (
+	adapter  = flag.String("adapter", "sim", "type of simulation: sim|socket|exec|docker")
+	loglevel = flag.Int("loglevel", 2, "verbosity of logs")
+)
+
+var (
+	defaultSkipCheck bool
+	waitPeerErrC     chan error
+	chunkSize        = 4096
+)
+
+var services = adapters.Services{
+	"streamer": NewStreamerService,
+}
+
+func init() {
+	flag.Parse()
+	// register the Delivery service which will run as a devp2p
+	// protocol when using the exec adapter
+	adapters.RegisterServices(services)
+
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*loglevel), log.StreamHandler(os.Stderr, log.TerminalFormat(false))))
+
+}
+
+// NewStreamerService
+func NewStreamerService(ctx *adapters.ServiceContext) (node.Service, error) {
+	id := ctx.Config.ID
+	addr := toAddr(id)
+	kad := network.NewKademlia(addr.Over(), network.NewKadParams())
+	store := stores[id]
+	db := storage.NewDBAPI(store.(*storage.LocalStore))
+	delivery := NewDelivery(kad, db)
+	deliveries[id] = delivery
+	r := NewRegistry(addr, delivery, store, defaultSkipCheck)
+	RegisterSwarmSyncerServer(r, db)
+	RegisterSwarmSyncerClient(r, db)
+	go func() {
+		waitPeerErrC <- waitForPeers(r, 1*time.Second, peerCount(id))
+	}()
+	return r, nil
+}
+
+func newStreamerTester(t *testing.T) (*p2ptest.ProtocolTester, *Registry, *storage.LocalStore, func(), error) {
+	// setup
+	addr := network.RandomAddr() // tested peers peer address
+	to := network.NewKademlia(addr.OAddr, network.NewKadParams())
+
+	// temp datadir
+	datadir, err := ioutil.TempDir("", "streamer")
+	if err != nil {
+		return nil, nil, nil, func() {}, err
+	}
+	teardown := func() {
+		os.RemoveAll(datadir)
+	}
+
+	localStore, err := storage.NewTestLocalStoreForAddr(datadir, addr.Over())
+	if err != nil {
+		return nil, nil, nil, teardown, err
+	}
+
+	db := storage.NewDBAPI(localStore)
+	delivery := NewDelivery(to, db)
+	streamer := NewRegistry(addr, delivery, localStore, defaultSkipCheck)
+	protocolTester := p2ptest.NewProtocolTester(t, network.NewNodeIDFromAddr(addr), 1, streamer.runProtocol)
+
+	err = waitForPeers(streamer, 1*time.Second, 1)
+	if err != nil {
+		return nil, nil, nil, nil, errors.New("timeout: peer is not created")
+	}
+
+	return protocolTester, streamer, localStore, teardown, nil
+}
+
+func waitForPeers(streamer *Registry, timeout time.Duration, expectedPeers int) error {
```

Comment:
This should be reworked using connection events, see https://github.com/ethersphere/go-ethereum/blob/98ba78c5218889f9159d8db1c5d5c29167cf7fc4/swarm/network/stream/testing/testing.go#L214
or 
 https://github.com/ethersphere/go-ethereum/blob/98ba78c5218889f9159d8db1c5d5c29167cf7fc4/swarm/network/simulations/discovery/discovery_test.go#L258